### PR TITLE
NuGetUpgrader: verify integrity of downloaded NuGet package

### PR DESCRIPTION
**Pull request by jamill**
Verify the integrity / signature of NuGet packages. The NuGet upgrader will
rely on the services provided by the NuGet client library to verify downloaded
NuGet packages.

Relies on the high level command `VerifyCommandRunner` exposed by NuGet client
library to actually perform NuGet package verification.

### CODE REVIEW

**Comment by wilbaker** [comment]
Before merging could you add some details to the PR description of what these changes do?

File: GVFS/GVFS.Common/NuGetUpgrader/NuGetFeed.cs [hunk 1]
```diff
@@ -128,6 +129,28 @@ public virtual async Task<string> DownloadPackageAsync(PackageIdentity packageId
             return downloadPath;
         }
 
+        public virtual bool VerifyPackage(string packagePath, string otherCertificateFingerprintToTrust)
+        {
+            List<string> certificateFingerprints = new List<string>();
+            certificateFingerprints.Add("3F9001EA83C560D712C24CF213C3D312CB3BFF51EE89435D3430BD06B5D0EECE");
+            if (otherCertificateFingerprintToTrust != null)
```
Comment (wilbaker) [inline GVFS/GVFS.Common/NuGetUpgrader/NuGetFeed.cs]: Where did the `"3F9001E...` fingerprint come from?

File: GVFS/GVFS.Common/NuGetUpgrader/NuGetFeed.cs [hunk 2]
```diff
@@ -128,6 +129,28 @@ public virtual async Task<string> DownloadPackageAsync(PackageIdentity packageId
...
+            List<string> certificateFingerprints = new List<string>();
+            certificateFingerprints.Add("3F9001EA83C560D712C24CF213C3D312CB3BFF51EE89435D3430BD06B5D0EECE");
+            if (otherCertificateFingerprintToTrust != null)
+            {
+                certificateFingerprints.Add(otherCertificateFingerprintToTrust);
+            }
+
+            VerifyArgs verifyArgs = new VerifyArgs()
+            {
+                Verifications = new VerifyArgs.Verification[] { VerifyArgs.Verification.All },
```
Comment (wilbaker) [inline GVFS/GVFS.Common/NuGetUpgrader/NuGetFeed.cs]: What does `VerifyArgs.Verification.All` mean?

File: GVFS/GVFS.Common/NuGetUpgrader/NuGetUpgrader.cs [hunk 3]
```diff
@@ -326,6 +337,17 @@ public override bool TryRunInstaller(InstallActionWrapper installActionWrapper,
...
+                    if (!this.noVerify)
+                    {
+                        if (!this.nuGetFeed.VerifyPackage(this.DownloadedPackagePath, this.nuGetUpgraderConfig.CertificateFingerprint))
+                        {
+                            error = "Package signature validation failed. Check the upgrade logs for more details.";
+                            activity.RelatedError(error);
+                            this.fileSystem.DeleteFile(this.DownloadedPackagePath);
+                            return false;
+                        }
+                    }
```
Comment (wilbaker) [inline GVFS/GVFS.Common/NuGetUpgrader/NuGetUpgrader.cs]: This code looks the same as the new code in `TryDownloadNewestVersion`, can it be moved to a helper method?

File: GVFS/GVFS.Common/NuGetUpgrader/NuGetUpgrader.cs [hunk 4]
```diff
@@ -460,6 +482,7 @@ public NuGetUpgraderConfig(ITracer tracer, LocalGVFSConfig localGVFSConfig)
             public string FeedUrl { get; private set; }
             public string PackageFeedName { get; private set; }
             public string FeedUrlForCredentials { get; private set; }
+            public string CertificateFingerprint { get; private set; }
```
Comment (wilbaker) [inline GVFS/GVFS.Common/NuGetUpgrader/NuGetUpgrader.cs]: Is this only for testing\debugging?  If so, will it be removed before merging to master? (on the surface it seems like a security risk to leave it in the production code)

File: GVFS/GVFS.Common/NuGetUpgrader/NuGetUpgrader.cs [hunk 5]
```diff
@@ -531,7 +554,17 @@ public bool TryLoad(out string error)
                     return false;
                 }
 
-                this.PackageFeedName = configValue;
+                this.PackageFeedName = configValue;
+
+                if (!this.localConfig.TryGetConfig("upgrade.certificatefingerprint", out configValue, out error))
+                {
+                    // This is option (and only for testing purposes)
```
Comment (wilbaker) [inline GVFS/GVFS.Common/NuGetUpgrader/NuGetUpgrader.cs]: `option` -> `optional`

File: GVFS/GVFS.Common/NuGetUpgrader/NuGetUpgrader.cs [hunk 6]
```diff
@@ -274,6 +274,17 @@ public override bool TryDownloadNewestVersion(out string errorMessage)
                 }
             }
 
+            if (!this.noVerify)
+            {
+                if (!this.nuGetFeed.VerifyPackage(this.DownloadedPackagePath, this.nuGetUpgraderConfig.CertificateFingerprint))
```
Comment (wilbaker) [inline GVFS/GVFS.Common/NuGetUpgrader/NuGetUpgrader.cs]: Add a mock for `VerifyPackage` to `private Mock<NuGetFeed> mockNuGetFeed` in `NuGetUpgraderTests` and you should be able to add a unit test the `TryDownloadNewestVersion` fails if package verification fails.

(If there are other unit tests code paths that hit `VerifyPackage` they can be updated as well)

File: GVFS/GVFS.Common/NuGetUpgrader/NuGetUpgrader.cs (hunk 4 above)
Comment (jamill) [reply GVFS/GVFS.Common/NuGetUpgrader/NuGetUpgrader.cs]: I do plan to remove it before merging. I don't think it actually increases our risk very much by itself, as it:

1) Must be configured in a file that requires admin privileges to access
2) The cert must be trusted by the machine, so it is not sufficient to just set this fingerprint.

However, we can consider this after the initial work here. This is contained in its own commit, so it should be easy to pull out.

File: GVFS/GVFS.Common/NuGetUpgrader/NuGetFeed.cs (hunk 2 above)
Comment (jamill) [reply GVFS/GVFS.Common/NuGetUpgrader/NuGetFeed.cs]: The documentation on this seems a bit murky... the [NuGet CLI](https://docs.microsoft.com/en-us/nuget/tools/cli-ref-verify) documentation indicates that "All possible verifications will be performed". It looks like the only verifications are "Signature" verifications at this point in time...

 See for [example](https://github.com/NuGet/NuGet.Client/blob/7ca842c4386e287ba456a115979a7946766509df/src/NuGet.Core/NuGet.Commands/VerifyCommand/VerifyCommandRunner.cs#L38)

File: GVFS/GVFS.Common/NuGetUpgrade/NuGetFeed.cs [hunk 7]
```diff
@@ -128,6 +129,21 @@ public virtual async Task<string> DownloadPackageAsync(PackageIdentity packageId
             return downloadPath;
         }
 
+        public virtual bool VerifyPackage(string packagePath)
+        {
+            VerifyArgs verifyArgs = new VerifyArgs()
+            {
+                Verifications = new VerifyArgs.Verification[] { VerifyArgs.Verification.All },
+                PackagePath = packagePath,
+                CertificateFingerprint = new List<string>() { "3F9001EA83C560D712C24CF213C3D312CB3BFF51EE89435D3430BD06B5D0EECE" },
```
Comment (wilbaker) [inline GVFS/GVFS.Common/NuGetUpgrade/NuGetFeed.cs]: As discussed offline, add a named constant for `3F9001EA83C560D712C24CF213C3D312CB3BFF51EE89435D3430BD06B5D0EECE`

File: GVFS/GVFS.UnitTests/Common/NuGetUpgrade/NuGetUpgraderTests.cs [hunk 8]
```diff
@@ -353,6 +356,72 @@ public void TrySetupToolsDirectoryFailsIfCreateToolsDirectoryFails()
...
+            string message;
+            List<IPackageSearchMetadata> availablePackages = new List<IPackageSearchMetadata>()
+            {
+                this.GeneratePackageSeachMetadata(new Version(CurrentVersion)),
+                this.GeneratePackageSeachMetadata(new Version(NewerVersion)),
+            };
+
+            IPackageSearchMetadata newestAvailableVersion = availablePackages.Last();
+
+            string downloadPath = "c:\\test_download_path";
```
Comment (wilbaker) [inline GVFS/GVFS.UnitTests/Common/NuGetUpgrade/NuGetUpgraderTests.cs]: Use `this.downloadDirectoryPath` instead.

**Pull request merged**